aea/package_manager/base.py

```diff
@@ -0,0 +1,426 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
```

Comment:
How much is new and how much is extension? This PR messes up concerns. It should really be two PRs. One just moving existing logic around, the other applying new logic!